File: bindings/node/examples/dealer-router-recv.js
// SPDX-License-Identifier: MPL-2.0

'use strict';

const zlink = require('../dist');

const ctx = new zlink.Context();
const router = new zlink.RouterSocket(ctx);
const dealer = new zlink.DealerSocket(ctx);

router.bind('inproc://example-dealer-router');
dealer.connect('inproc://example-dealer-router');
dealer.send(zlink.Message.copyOf('hello'));

const request = router.recv();
console.log(request.parts[0].toString());
router.sendTo(request.routingId, zlink.Message.copyOf('world'));

const response = dealer.recv();
console.log(response.parts[0].toString());

dealer.close();
router.close();
ctx.close();

File: bindings/python/examples/dealer_router_recv.py
import os, sys
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import zlink
from sample_common import tcp_endpoint, wait_connected


def main():
    port, endpoint = tcp_endpoint()
    with zlink.Context() as ctx:
        with zlink.RouterSocket(ctx) as router:
            with zlink.DealerSocket(ctx) as dealer:
                with router.open_monitor(zlink.MonitorEvent.CONNECTION_READY_CHANGED) as rtr_mon:
                    with dealer.open_monitor(zlink.MonitorEvent.CONNECTION_READY_CHANGED) as dlr_mon:
                        dealer.set_routing_id(b"CLIENT")
                        router.bind(endpoint)
                        dealer.connect(endpoint)
                        wait_connected(rtr_mon, dlr_mon)

                dealer.send(b"ping")
                with router.recv() as request:
                    router.send(b"pong", routing_id=request.routing_id)

                with dealer.recv() as response:
                    data = response.to_bytes_list()[0].decode("utf-8")
                    print(f'[dealer-router/recv] send: "ping" \u2192 recv: "{data}"')


if __name__ == "__main__":
    main()
